### PR TITLE
SALTO-5879: Okta: Add support for addition and removal of `BrandTheme`

### PR DESCRIPTION
**Pull request by Nurdok**
Add support for addition and removal of `BrandTheme`.

Addition: `BrandTheme`s are created automatically with default values. This PR allows to "add" a custom theme (alongside its Brand) by modifying the automatically created theme instead.

Removal: `BrandTheme`s are automatically removed when their `Brand`s are removed. This PR implements removal by just verifying that they are removed (and disallowing removal if the `Brand` wasn't removed as well).

---

_Additional context for reviewer_
Removal handling is similar to #5594

---
_Release Notes_: 
_Okta adapter_:
- Added support for addition and removal of `BrandTheme`.

---
_User Notifications_: 
None

### CODE REVIEW

File: packages/okta-adapter/src/change_validators/brand_theme_removal.ts [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+  const removeInstanceChanges = changes.filter(isInstanceChange).filter(isRemovalChange).map(getChangeData)
+
+  const removedBrandThemeInstances = removeInstanceChanges.filter(
+    instance => instance.elemID.typeName === BRAND_THEME_TYPE_NAME,
+  )
+
+  const removedNames = new Set(removeInstanceChanges.map(instance => instance.elemID.getFullName()))
+
+  return removedBrandThemeInstances
+    .filter(brandTheme => !removedNames.has(getParent(brandTheme).elemID.getFullName()))
```
Comment (shir-reifenberg) [inline packages/okta-adapter/src/change_validators/brand_theme_removal.ts:40]: `getParent` might throw, can you please wrap it with try-catch or use `getParents` instead?

File: packages/okta-adapter/src/config.ts [hunk 2]
```diff
@@ -953,6 +953,15 @@ const DEFAULT_TYPE_CUSTOMIZATIONS: OktaSwaggerApiConfig['types'] = {
...
     },
     deployRequests: {
+      add: {
+        url: '/api/v1/brands/{brandId}/themes/{themeId}',
+        method: 'put',
+        urlParamsToFields: {
+          brandId: '_parent.0.id',
+          themeId: 'id',
+        },
+        fieldsToIgnore: ['id', 'logo', 'favicon', '_links'],
```
Comment (shir-reifenberg) [inline packages/okta-adapter/src/config.ts:963]: why do we need those fields to ignore? those are deleted already during fetch, no ?

File: packages/okta-adapter/src/filters/brand_theme_addition.ts [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+}
+
+const deployBrandThemeAddition = async (
+  change: Change<InstanceElement>,
+  client: OktaClient,
+  apiDefinitions: OktaSwaggerApiConfig,
+): Promise<void> => {
+  const instance = getChangeData(change)
+  // eslint-disable-next-line no-underscore-dangle
+  const brandId = instance.annotations?._parent?.[0]?.id
```
Comment (shir-reifenberg) [inline packages/okta-adapter/src/filters/brand_theme_addition.ts:48]: please use `getParent` / `getParents`

File: packages/okta-adapter/src/filters/brand_theme_addition.ts [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+  const themeEntries = (
+    await client.get({
+      url: `/api/v1/brands/${brandId}/themes`,
+    })
+  ).data
+  if (_.isArray(themeEntries) && themeEntries.length === 1 && _.isString(themeEntries[0].id)) {
+    return themeEntries[0].id
+  }
+  log.error(`Received unexpected result for brand theme: ${inspectValue(themeEntries)}`)
+  throw new Error(`Could not find BrandTheme with the provided brandId ${brandId}`)
```
Comment (shir-reifenberg) [inline packages/okta-adapter/src/filters/brand_theme_addition.ts:38]: This error will be displayed to the user, no? I'm not sure we should include the `brandId` here..

File: packages/okta-adapter/src/filters/brand_theme_removal.ts [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+import _ from 'lodash'
+import { getParents } from '@salto-io/adapter-utils'
+import { isInstanceChange, getChangeData, isRemovalChange } from '@salto-io/adapter-api'
+import { client as clientUtils } from '@salto-io/adapter-components'
+import { FilterCreator } from '../filter'
+import { BRAND_THEME_TYPE_NAME } from '../constants'
+import { deployChanges } from '../deployment'
+import OktaClient from '../client/client'
+
+const verifyBrandThemeIsDeleted = async (
```
Comment (shir-reifenberg) [inline packages/okta-adapter/src/filters/brand_theme_removal.ts:25]: I wonder if we should consolidate with `verifyProfileMappingIsDeleted` someway, WDYT? (if it's not much effort, cause we should move the the new infra in the near future..)
Also, let's add some logs doe the unexpected results ?

File: packages/okta-adapter/test/change_validators/brand_theme_removal.test.ts [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+  ObjectType,
+  ElemID,
+  InstanceElement,
+  ReferenceExpression,
+  CORE_ANNOTATIONS,
+} from '@salto-io/adapter-api'
+import { brandThemeRemovalValidator } from '../../src/change_validators/brand_theme_removal'
+import { OKTA, BRAND_THEME_TYPE_NAME, BRAND_TYPE_NAME } from '../../src/constants'
+
+describe('profileMappingRemovalValidator', () => {
```
Comment (shir-reifenberg) [inline packages/okta-adapter/test/change_validators/brand_theme_removal.test.ts:27]: please change name :)

File: packages/okta-adapter/test/filters/brand_theme_addition.test.ts [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+
+  beforeEach(() => {
+    jest.clearAllMocks()
+    const { client: cli, connection } = mockClient()
+    mockConnection = connection
+    client = cli
+    filter = brandThemeAdditionFilter(getFilterParams({ client })) as typeof filter
+  })
+
+  describe('deploy', () => {
```
Comment (shir-reifenberg) [inline packages/okta-adapter/test/filters/brand_theme_addition.test.ts:49]: can we please add tests for the failure scenarios ?

**Review by shir-reifenberg — COMMENTED** [review]
Looks great! 😎 Added some comments.
Also, looks like coverage decreased, can you please add the missing tests ? (test file for removal changes, cover the logo.ts change in test..)

File: packages/okta-adapter/src/change_validators/brand_theme_removal.ts (hunk 1 above)
Comment (Nurdok) [reply packages/okta-adapter/src/change_validators/brand_theme_removal.ts:40]: Done

File: packages/okta-adapter/src/config.ts (hunk 2 above)
Comment (Nurdok) [reply packages/okta-adapter/src/config.ts:963]: Done.

File: packages/okta-adapter/src/filters/brand_theme_addition.ts (hunk 4 above)
Comment (Nurdok) [reply packages/okta-adapter/src/filters/brand_theme_addition.ts:38]: Moved to the log line above to still keep this information for debugging.

File: packages/okta-adapter/test/change_validators/brand_theme_removal.test.ts (hunk 6 above)
Comment (Nurdok) [reply packages/okta-adapter/test/change_validators/brand_theme_removal.test.ts:27]: Done

File: packages/okta-adapter/src/filters/brand_theme_addition.ts (hunk 3 above)
Comment (Nurdok) [reply packages/okta-adapter/src/filters/brand_theme_addition.ts:48]: Done.

File: packages/okta-adapter/src/filters/brand_theme_addition.ts [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+  throw new Error('Could not find BrandTheme with the provided brandId')
+}
+
+const deployBrandThemeAddition = async (
+  change: Change<InstanceElement>,
+  client: OktaClient,
+  apiDefinitions: OktaSwaggerApiConfig,
+): Promise<void> => {
+  const instance = getChangeData(change)
+  // eslint-disable-next-line no-underscore-dangle
```
Comment (shir-reifenberg) [inline packages/okta-adapter/src/filters/brand_theme_addition.ts:47]: needed ?

File: packages/okta-adapter/src/filters/brand_theme_removal.ts (hunk 5 above)
Comment (shir-reifenberg) [reply packages/okta-adapter/src/filters/brand_theme_removal.ts:28]: answered there as well :)

**Review by shir-reifenberg — APPROVED** [review]
LGTM with one nit 😎

**Pull request merged**